radicale/storage.py

```diff
@@ -687,53 +862,87 @@ def delete(self, href=None):
-            items.append(self.get(href).item)
-        time_end = datetime.datetime.now()
-        self.logger.info(
-            "Collection read %d items in %s sec from %s", len(items),
-            (time_end - time_begin).total_seconds(), self._filesystem_path)
+        # serialize collection
         if self.get_meta("tag") == "VCALENDAR":
-            collection = vobject.iCalendar()
-            for item in items:
-                for content in ("vevent", "vtodo", "vjournal"):
-                    if content in item.contents:
-                        for item_part in getattr(item, "%s_list" % content):
-                            collection.add(item_part)
-                        break
-            return collection.serialize()
+            vtimezones = ""
+            included_tzids = set()
+            vtimezone = []
+            tzid = None
+            components = ""
+            for item in self.get_all():
+                depth = 0
+                for line in item.serialize().split("\r\n"):
+                    if line.startswith("BEGIN:"):
+                        depth += 1
+                    if depth == 2 and line == "BEGIN:VTIMEZONE":
+                        vtimezone.append(line)
+                    elif vtimezone:
+                        vtimezone.append(line)
+                        if depth == 2 and line.startswith("TZID:"):
+                            tzid = line[len("TZID:"):]
+                        elif depth == 2 and line.startswith("END:"):
+                            if tzid is None or tzid not in included_tzids:
+                                if vtimezones:
+                                    vtimezones += "\r\n"
+                                vtimezones += "\r\n".join(vtimezone)
+                                included_tzids.add(tzid)
+                            vtimezone.clear()
+                            tzid = None
+                    elif depth >= 2:
+                        if components:
+                            components += "\r\n"
+                        components += line
+                    if line.startswith("END:"):
+                        depth -= 1
+            return "\r\n".join(filter(bool, (
+                "BEGIN:VCALENDAR",
+                "VERSION:2.0",
+                "PRODID:-//PYVOBJECT//NONSGML Version 1//EN",
+                vtimezones,
+                components,
+                "END:VCALENDAR")))
         elif self.get_meta("tag") == "VADDRESSBOOK":
-            return "".join([item.serialize() for item in items])
+            return "".join(map(lambda x: x.serialize(), self.get_all()))
         return ""
 
+    @property
+    def etag(self):
+        # reuse cached value if the storage is read-only
+        if self._writer or self._etag is None:
+            etag = md5()
+            for item in self.get_all():
+                etag.update((item.href + "/" + item.etag).encode("utf-8"))
+            self._etag = '"%s"' % etag.hexdigest()
```

Comment:
> Using string-concatenation

I used this method, because it's also used at other places in the code. I don't think that it matters.

> should be as fast.

This is only done once, speed is not a concern.